### PR TITLE
Added baseUrl to aid file resolution, added subset config for Iotas 

### PR DESCRIPTION
**Pull request by jsoverson**
This PR:
- Removes instances of changing the working directory in favor of tracking the baseUrl per imported configuration
- Adds a subset configuration for included Iotas to differentiate them from root applications that have features like transports etc.

### CODE REVIEW

**Comment by pkedy** [comment]
> Removes instances of changing the working directory in favor of tracking the baseUrl per imported configuration

How did you account for actions, etc loading files?

**Comment by jsoverson** [comment]
>  How did you account for actions, etc loading files?

Do you have any examples or tests to exercise that logic? Without tests I could only use the existing examples (url shortener & nanochat) as a guide for what to account for.

**Comment by jsoverson** [comment]
@pkedy I redid this branch off of main to separate it from the config-dx work.

File: pkg/config/paths.go [hunk 1]
```diff
@@ -0,0 +1,46 @@
...
+	if len(base) > 0 {
+		baseUrl = GetBase(&base[0])
+	} else {
+		baseUrl = GetBase(nil)
+	}
+	if strings.HasPrefix(url, ".") {
+		normalUrl, err = URL.Parse(fmt.Sprintf("file:///%s", path.Join(baseUrl, url)))
+	} else if strings.HasPrefix(url, "/") {
+		normalUrl, err = URL.Parse(fmt.Sprintf("file:///%s", url))
+	} else {
```
Comment (pkedy) [inline pkg/config/paths.go:38]: Would it make sense for another case to look `http://` or `https://` and the else case using `path.Join`? When working with local files, I personally dislike having to remember to prefix with `./`. I get the "modules"/ESM approach though.

File: pkg/config/paths.go [hunk 2]
```diff
@@ -0,0 +1,46 @@
...
+func NormalizeUrl(url string, base ...string) (string, error) {
+	var normalUrl *URL.URL
+	var err error
+	var baseUrl string
+	if len(base) > 0 {
+		baseUrl = GetBase(&base[0])
+	} else {
+		baseUrl = GetBase(nil)
+	}
+	if strings.HasPrefix(url, ".") {
```
Comment (pkedy) [inline pkg/config/paths.go:34]: If `url` contains `..`, is it possible to exploit the FS? Just something to think about.

File: pkg/engine/engine.go [hunk 3]
```diff
@@ -561,46 +567,26 @@ func Start(info *Info) error {
...
-	}
-	for instanceID, include := range busConfig.Includes {
-		if oci.IsImageReference(include.Ref) {
-			return errors.New("references not currently supported")
-		}
-		log.Info("Loading include", "name", instanceID, "ref", include.Ref)
-		path := filepath.Join(dir, include.Ref)
-		var busFile, parentDir string
-		info, err := os.Stat(path)
+	for name, include := range busConfig.Includes {
```
Comment (pkedy) [inline pkg/engine/engine.go:570]: I think this was my mistake but this construct is supposed to be called "uses" (or maybe import). I see "includes" as more of a file inclusion for breaking up the YAML into multiple files. Not as a Iota boundary.

File: pkg/runtime/apex.axdl [hunk 4]
```diff
@@ -23,19 +23,19 @@ type ResourcesConfig {
...
+  main: string?
   "The interface definition."
-  spec: string
+  spec: string?
   "Package specifies the contents of the Iota OCI image."
   package: Package?
   "Resources are externally configured sources and receivers of data (DB, REST endpoint)."
   resources: [string]?
-  "Other Iotas that this Iota depends on using."
+  "Imported Iota dependencies."
```
Comment (pkedy) [inline pkg/runtime/apex.axdl:38]: As I mentioned above, perhaps we use a different term here (e.g. `uses` or `imports`). `uses` is my preference... I guess. :)

File: pkg/runtime/iota.go [hunk 5]
```diff
@@ -0,0 +1,72 @@
...
+
+	"github.com/go-logr/logr"
+	"github.com/nanobus/nanobus/pkg/config"
+
+	"github.com/nanobus/nanobus/pkg/oci"
+)
+
+type iotaKey struct{}
+
+func FromContext(ctx context.Context) IotaConfig {
```
Comment (pkedy) [inline pkg/runtime/iota.go:25]: How will this be used? I was thinking something along these lines for how actions under an iota can resolve files, but wasn't thinking the entire config struct.

**Review by pkedy — COMMENTED** [review]
My only reservation is requiring a `./` prefix when working with local files. See comments for proposal.

File: pkg/config/paths.go (hunk 1 above)
Comment (jsoverson) [reply pkg/config/paths.go:38]: good point. I'll invert the logic and parse it as a URL and if there's an error or the scheme is `''` then I'll treat it as a local file.

File: pkg/runtime/apex.axdl (hunk 4 above)
Comment (pkedy) [reply pkg/runtime/apex.axdl:38]: On second thought, I'm using `uses` for something else in the TypeScript stuff (e.g. `app.uses(RestModuleV1(...))`). Maybe we want `imports`.

**Comment by jsoverson** [comment]
I inverted the path resolution logic and changed includes->imports. During the latter change I got stuck on some bad error messages from the yaml decoding which led to some engine changes for testability and improved config errors, e.g.

Instead of:

```
could not load configuration: yaml: unmarshal errors:\\n  line 15: field wtf not found in type runtime.alias
```

They're now:

```
could not load configuration: line 15: field 'wtf' not expected
```

File: pkg/config/yaml.go [hunk 6]
```diff
@@ -1,19 +1,39 @@
...
+			// Then extract the valuable data from the errors and make them more useful.
+			for _, invalidField := range typeError.Errors {
+				found := reYamlError.FindStringSubmatch(invalidField)
+				if len(found) == 3 {
+					line := found[1]
+					field := found[2]
+					errs = append(errs, fmt.Sprintf("line %s: field '%s' not expected", line, field))
+				}
+			}
+			fmt.Println(strings.Join(errs, ", "))
```
Comment (pkedy) [inline pkg/config/yaml.go:34]: Is this `fmt.Println` left over debugging code? Would use the returned error to contain the message we want and log it from the caller.

**Review by pkedy — APPROVED** [review]
LGTM (see one comment)

File: pkg/config/yaml.go (hunk 6 above)
Comment (jsoverson) [reply pkg/config/yaml.go:34]: Yes, thanks for catching it. There are tests to assert no errant output on successful invoke, but not for failures.

**Review by pkedy — APPROVED** [review]

**Pull request merged**